kubernetes/kserve/kf_request_json/v2/bert/README.md

```diff
@@ -0,0 +1,115 @@
+
+```
+python Download_Transformer_models.py
+```
+ 
+ Generate mar file
+ 
+```bash
+torch-model-archiver --model-name BERTSeqClassification --version 1.0 \
+--serialized-file Transformer_model/pytorch_model.bin \
+--handler ./Transformer_handler_generalized_v2.py \
+--extra-files "Transformer_model/config.json,./setup_config.json,./Seq_classification_artifacts/index_to_name.json,./Transformer_handler_generalized.py"
+```
+
+The command will create `BERTSeqClassification.mar` file in current directory
+
+Move the mar file to model-store(`/mnt/models/model-store`) folder 
+
+and use the following config properties (`/mnt/models/config`)
+
+```
+inference_address=http://0.0.0.0:8085
+management_address=http://0.0.0.0:8085
+metrics_address=http://0.0.0.0:8082
+enable_envvars_config=true
+install_py_dep_per_model=true
+enable_metrics_api=true
+service_envelope=kservev2
+metrics_format=prometheus
+NUM_WORKERS=1
+number_of_netty_threads=4
+job_queue_size=10
+model_store=/mnt/models/model-store
+model_snapshot={"name":"startup.cfg","modelCount":1,"models":{"BERTSeqClassification":{"1.0":{"defaultVersion":true,"marName":"BERTSeqClassification.mar","minWorkers":1,"maxWorkers":5,"batchSize":1,"maxBatchDelay":5000,"responseTimeout":120}}}}
+```
+
+## Preparing input
+
+Use [bert_bytes_v2.json](bert_bytes_v2.json) or [bert_tensor_v2](bert_tensor_v2.json).
+
+For new sample text, follow the instructions below
+
+For bytes input, use [tobytes](tobytes.py) utility. 
+
+```
+python tobytes.py --input_text "this year business is good"
+```
+
+For tensor input, use [bert_tokenizer](bert_tokenizer.py) utility
+
+```
+python bert_tokenizer.py --input_text "this year business is good"
+```
+
+
+## Deploying the model in local machine
+
+Start TorchServe
+
+```
+torchserve --start --ts-config /mnt/models/config/config.properties --ncs
+```
+
+To test locally, clone TorchServe and move to the following folder `kubernetes/kserve/kserve_wrapper`
+
+Start Kserve
+
+```
+python __main__.py
+```
+
+## Sample request and response for bytes input
+
+Navigate to `kubernetes/kserve/kf_request_json/v2/bert`
+
+Run the following command
+
+```bash
+curl -v -H "ContentType: application/json" http://localhost:8080/v2/models/BERTSeqClassification/infer -d @./bert_bytes_v2.json
+```
+
+Expected Output
+
+```bash
+{"id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298", "model_name": "BERTSeqClassification", "model_version": "1.0", "outputs": [{"name": "predict", "shape": [], "datatype": "BYTES", "data": ["Not Accepted"]}]}
+```
+
+
+## Sample request and response for tensor input
+
+
+Run the following command
+
+```
+curl -v -H "ContentType: application/json" http://localhost:8080/v2/models/BERTSeqClassification/infer -d @./bert_tensor_v2.json
+```
+
+Expected output
+```bash
+{"id": "33abc661-7265-42fc-b7d9-44e5f79a7a67", "model_name": "BERTSeqClassification", "model_version": "1.0", "outputs": [{"name": "predict", "shape": [], "datatype": "BYTES", "data": ["Not Accepted"]}]}
+```
```

Comment:
I extended the `get_insights` method and the explain for bert is working with bytes input - https://github.com/shrinath-suresh/kserve/commit/0e64e897779fde0efdfe9aa118dcaba223d0c182. 

For infer endpoint there is only one input - input_text. However, for the explain endpoint, there are two inputs  (input_text and target). 

The input structure for torchserve native and kserve v2 are completely different. Torchserve native takes json input, hence it is really easy to send the input like [this](https://github.com/pytorch/serve/blob/master/examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text_captum_input.txt)

However, the input structure for kserve is like [this](https://github.com/kserve/kserve/blob/master/docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/bytes/bert_v2.json)

We can take two different approaches

1. send the json string input in bytes data field. Something like this

```

{
  "id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298",
  "inputs": [{
    "name": "4b7c7d4a-51e4-43c8-af61-04639f6ef4bc",
    "shape": -1,
    "datatype": "BYTES",
    "data": '{"text":"Bloomberg has decided to publish a new report on the global economy.", "target":1} '
  }]
}
```

At the kserve handler - we will receive the input as Bytes. We need to decode and then parse the dict to fetch the values. We need to change the infer part accordingly (unless we decide to keep two different jsons for infer and explain)

2. Another option is to send it as individual inputs (since `inputs` tag is already an array)

```
{
  "id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298",
  "inputs": [{
    "name": "4b7c7d4a-51e4-43c8-af61-04639f6ef4bc",
    "shape": -1,
    "datatype": "BYTES",
    "data": 'Bloomberg has decided to publish a new report on the global economy.'
  },
{
    "name": "4b7c7d4a-51e4-43c8-af61-04639f6ef4bc",
    "shape": 1,
    "datatype": "INT64",
    "data": 1
  },]
}
```
As per v2 spec, this is how the request should be structured as there are multiple inputs. However, this may not work well on the batched input scenarios. Imagine, there are multiple inputs at the same time and it is difficult to map between the text and target.

I am currently trying approach 1. Will raise a PR on kserve side once its tested. Please let me know your thoughts as well.

Reply:
Thanks @shrinath-suresh,  approach 1 sounds good to me as well.

Reply:
@shrinath-suresh Is the PR on kserve side ready?

Reply:
Yes @lxning . All the PRs on the kserve side has been already merged.

https://github.com/kserve/kserve/pull/2043
https://github.com/kserve/kserve/pull/2035

We can go ahead and merge this PR.